internal/config/variables.go

```diff
@@ -0,0 +1,420 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/ext/typeexpr"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+)
+
+// A consistent detail message for all "not a valid identifier" diagnostics.
+const badIdentifierDetail = "A name must start with a letter or underscore and may contain only letters, digits, underscores, and dashes."
+
+// VariableAssignments contain the values from the job, from the server/VCS
+// repo, and default values from the waypoint.hcl. These are to
+// sort precedence and add the final variable values to the EvalContext
+type VariableAssignment struct {
+	Value  cty.Value
+	Source string
+	Expr   hcl.Expression
+	// The location of the variable value if the value was provided
+	// from a file
+	Range hcl.Range
+}
+
+// TODO krantzinator: this goes somewhere else
+type Variable struct {
+	Name string
+
+	// Values contains possible values for the variable; The last value set
+	// from these will be the one used. If none is set; an error will be
+	// returned by Value().
+	Values []VariableAssignment
+
+	// Cty Type of the variable. If the default value or a collected value is
+	// not of this type nor can be converted to this type an error diagnostic
+	// will show up. This allows us to assume that values are valid.
+	//
+	// When a default value - and no type - is passed into the variable
+	// declaration, the type of the default variable will be used.
+	Type cty.Type
+
+	// Description of the variable
+	Description string
+
+	// The location of the variable definition block in the waypoint.hcl
+	Range hcl.Range
+}
+
+type hclVariable struct {
+	Name        string         `hcl:",label"`
+	Default     cty.Value      `hcl:"default,optional"`
+	Type        hcl.Expression `hcl:"type,optional"`
+	Description string         `hcl:"description,optional"`
+}
+
+// Variables are used when parsing the Config, to set default values from
+// the waypoint.hcl and bring in the values from the job and the server/VCS
+// for eventual precedence sorting and setting on the EvalContext
+type Variables map[string]*Variable
+
+// TODO krantzinator - use implied body scheme instead?
+var variableBlockSchema = &hcl.BodySchema{
+	Attributes: []hcl.AttributeSchema{
+		{
+			Name: "default",
+		},
+		{
+			Name: "type",
+		},
+		{
+			Name: "description",
+		},
+	},
+}
+
+func (c *Config) DecodeVariableBlocks(variables *Variables) hcl.Diagnostics {
```

Comment:
This feels like it shouldn't be a separate thing and should always just happen on config parse. Any reason to make this separate that I'm missing? 

My thought would it happens always on `Load` and there is either a struct field or a function to return the `*Variables` type.

Reply:
This is now in Config and makes a call to the variables package for individual block parsing. It returns a `map[string]*variables.Variable`.